### PR TITLE
Added Edit this Page Button

### LINKED ISSUE #50: Add "edit this page" banner to view source content of a page in GitHub
# Summary

Add a banner, button, or something else distinct titled "edit this page", which when clicked, takes the user to the source material in the source forge of choice (e.g. an AsciiDoc page on GitHub.com).

# Background

**Is the new feature related to a problem? Describe the problem**: All the content on the site is open source, but it is not clear or intuitive that edits are welcome or accepted. A call-out makes it easier to attract casual contributors by people flying by the website.

**What does the new feature look like to you?**: This banner or button should appear near the top and bottom of a page. It could also be fixed in the viewport as the user scrolls, but this needs careful consideration for mobile. More design thinking could be done on how best to design this feature. 

**Describe any alternatives considered**: We can always manually link the source material in a page, but this runs a high risk of becoming outdated if the page title or URL ever changes. Something automated is best so it will keep up with the document metadata instead of requiring an editor to manually fix a link.

# Details

I learned about the possibility to do this while reading the [Hugo v0.92.0 changelog](https://github.com/gohugoio/hugo/releases/tag/v0.92.0), which talks about `.Path` and how this is changing to a canonical content path in Hugo v0.93. So, given the functionality is there and all of the content is public here on GitHub, we should make it easier for contributions to happen.

This might require a new config value, maybe `edit-prefix` which would look something like `https://github.com/unicef/inventory/tree/main/content`. Pseudo code below:

```html
<a href="{{ $.site.params.edit-prefix }}/{{ .Path }}">edit this page</a>
```

# Outcome

Encourage contributions to content source material

### PR DESCRIPTION
**Pull request by Neha9849**
**What i did in this PR-**

- I added an Edit this Page Button in between the Breadcrumbs and Title in the content pages.

- I added a edit-prefix param in the config file which points to the content folder of the github repository.

 Fixes #50
 This PR is not yet ready to be merged, i raised this PR so that i can get help.

**Problem i am facing-**
When i passed the href attribute to the hyperlink, it is getting encoded and i am getting a 404 error.
Actual Outcome-
https://github.com/unicef/inventory-hugo-theme/blob/main/exampleSite/content/installation%5c_index.en.md
Expected Outcome-
https://github.com/unicef/inventory-hugo-theme/blob/main/exampleSite/content/installation/_index.en.md
 
I thought of decoding URL either by using Javascript `decodeURIComponent() `function or manually replacing %5c with / using replace function.
But i think there could be some better solution.
@jwflory @Idadelveloper can you look into this :)

### CODE REVIEW

**Comment by Neha9849** [comment]
Also I tried eliminate the extra variable `edit-prefix` which we have to add manually in all the downstreams by using the github repository link in the config file under` social` param but this caused an issue as the folder struture of this repository is different 
from the downstreams. As the content folder here is under exampleSite and in the downstreams it is under Root.

**Comment by jwflory** [comment]
Rebased on latest `main`, see commit 981863b5cc70d26545348cdaba50382a0947a1ea.

File: layouts/partials/default.html
Comment (jwflory) [inline layouts/partials/default.html:28]: This line causes a warning in the Hugo server:

`WARN 2022/04/05 21:58:03 .File.Path on zero object. Wrap it in if or with: {{ with .File }}{{ .Path }}{{ end }}`

Could you refactor this to resolve the new warning?

**Review by jwflory — CHANGES_REQUESTED** [review]
Nice work @Neha9849, this is great! 🙌🏻 I ran these changes locally and it worked as expected. I was able to traverse the example site and open the source page on GitHub.

The design is ready, but there are two changes to make for this to be ready:

1. The Hugo server now gives a warning when building the site with uncertain syntax for a zero object. See in-line comment for details.
2. Sometimes a maintainer may want to opt a page out of the edit button. Let's make it hidden on request by a front-matter boolean condition. This can be a simple check on a boolean condition in the page, e.g. `if (hide_edit_button != true) then (<<edit button logic>>)`

Does this feedback make sense? Once these two parts are implemented, this is ready to merge.

Note that I rebased the feature branch while working on this Pull Request, so the commit history has changed in the remote. You will need to check out the branch and run `git fetch` and `git reset --hard <remote>/edit-this-page-btn` to start fresh for new changes. Make sure any other work is committed on other branches before running these commands.

File: layouts/partials/default.html
Comment (Neha9849) [reply layouts/partials/default.html:28]: It's just a little change i will fix it!

**Comment by Neha9849** [comment]
@jwflory I am not getting the expected outcome 🥲 I described the problem above, can you please look into it.
I guess its something to do with the browser or OS not sure though.

> Sometimes a maintainer may want to opt a page out of the edit button. Let's make it hidden on request by a front-matter boolean condition. This can be a simple check on a boolean condition in the page, e.g. if (hide_edit_button != true) then (<<edit button logic>>)

Yeah it makes sense! I will make this change.

**Comment by Neha9849** [comment]
@jwflory I have made the requested changes.
The edit button gets displayed by default. If the maintainer don't want this, than the button gets hidden by setting a frontmatter variable `hide_edit_btn` to `true`.
I also did some slight padding changes to remove white space.

**Comment by jwflory** [comment]
Hi @Neha9849, thanks for your patience on the follow-up! 🙏🏻 A lot of new changes were merged to repo since this was opened. I am in the process of rebasing your Pull Request on the latest changes in `main`. I suggest mirroring your fork's `main` branch to the same `HEAD` as the `main` branch here. You can do so with a `git reset --hard upstream/main` when checked out on your local `main` branch.

Note that I will forcibly push to the `edit-this-page-btn` branch and this will require you to do a hard reset in your local branch to the remote branch. I will finish up the rebase and provide more guidance and final review after.

File: layouts/partials/default.html
Comment (jwflory) [inline layouts/partials/default.html:45]: This line seems to be misplaced. Note the Bootstrap class change from `p-4` to `p-3` and the same `id` value as line 27. I'm not sure which one is correct or if the change was accidental.

File: layouts/partials/default.html
Comment (jwflory) [inline layouts/partials/default.html:39]: Seems like you will need to wrap the `<a>` element inside of the `<button>`.

**Review by jwflory — CHANGES_REQUESTED** [review]
Hey Neha, changes look good and the opt-out behavior is what we needed. However, there are a couple more details to work out. The HTML is not validating (see screenshot below). I used the [W3 HTML Checker tool](https://validator.w3.org/nu/) to get this feedback. See specific details in-line below.

Once these are resolved, this feature should be ready to merge.

Screenshot of invalid HTML errors and warnings from the HTML validator service.

File: layouts/partials/default.html
Comment (Neha9849) [reply layouts/partials/default.html]: It was accidental. I removed it🤞

File: layouts/partials/default.html
Comment (Neha9849) [reply layouts/partials/default.html]: I did this to make the whole button a link. May be this isn't the good practice. So,I used a form to make the button a link which redirects to the desired page.

**Comment by Neha9849** [comment]
@jwflory I have made the requested changes.
However before merging this one,
we should add a `editPrefix` variable in the config file of inventory repository.

File: exampleSite/config.yaml
Comment (jwflory) [inline exampleSite/config.yaml:121]: `editPrefix` appears three times in the config file. Please reduce it to a single invocation. I suggest keeping it towards the top of the `params`, separate from the `brand` section.

File: exampleSite/config.yaml
Comment (jwflory) [inline exampleSite/config.yaml:99]: Not sure why the logo section was changed. Please revert this change to keep the scope narrow and clear.

BTW, remind me to show you how to do git patch commits sometimes with `git add --patch`. It is a super handy way to avoid committing too many changes by mistake.

**Review by jwflory — CHANGES_REQUESTED** [review]
Hey Neha, thanks for your steady work here. 🙏🏻  This PR is nearly ready! However, there are two more things to address:

1. **HTML 5 validation**: There are still HTML validation errors. It appears one of the colors is not being loaded correctly. Could you look into this more? _Hint_: The [W3 Validator](https://validator.w3.org/) is a useful tool. You can paste HTML from your browser's Developer Tools for quick validation while the Hugo server is running.
1. **Reduce commit footprint**: There are additional, unrelated changes committed to this Pull Request. We want to keep the PR scope lean and focused, and other minor changes should not be committed. This speeds up the review process for future changes too.

Screenshot of HTML 5 validation errors from the W3 HTML Validator tool.

Once these remaining tasks are cleared, we should be on track to merge! 🎬

File: exampleSite/config.yaml
Comment (Neha9849) [reply exampleSite/config.yaml]: I have this fixed this and added the editPrefix at the top of `params `  section.

Comment (Neha9849) [reply exampleSite/config.yaml]: > 1. **HTML 5 validation**: There are still HTML validation errors. It appears one of the colors is not being loaded correctly. Could you look into this more? _Hint_: The [W3 Validator](https://validator.w3.org/) is a useful tool. You can paste HTML from your browser's Developer Tools for quick validation while the Hugo server is running.

I tried pasting the html source code of a page and it showed something different from the above. I am unable to figure out the exact place where this errors occured as i dont know the code you pasted in W3 validator. The below three includes a anchor tag containing UNICEF, this change is not something i did. Could please provide the code you pasted, so that i can figure where it's going wrong.

> 2. **Reduce commit footprint**: There are additional, unrelated changes committed to this Pull Request. We want to keep the PR scope lean and focused, and other minor changes should not be committed. This speeds up the review process for future changes too.
This PR only includes changes which are relevant.
The commit history include previous commits. I exactly don't know why this happened, may be this is because i did some changes in main and they got into this branch while doing merge.
I tried doing a git rebase and squashing all the commits but i got some conflits. so i aborted it.

File: exampleSite/config.yaml
Comment (jwflory) [reply exampleSite/config.yaml]: Looks good! 👍🏻

**Comment by jwflory** [comment]
> **Problem i am facing-** When i passed the href attribute to the hyperlink, it is getting encoded and i am getting a 404 error. Actual Outcome- https://github.com/unicef/inventory-hugo-theme/blob/main/exampleSite/content/installation%5c_index.en.md Expected Outcome- https://github.com/unicef/inventory-hugo-theme/blob/main/exampleSite/content/installation/_index.en.md
> 
> I thought of decoding URL either by using Javascript `decodeURIComponent() `function or manually replacing %5c with / using replace function. But i think there could be some better solution.

Hey @Neha9849, are you still facing this? I wasn't able to replicate this in my testing. The edit button works even for index pages. The string was escaped correctly.

File: exampleSite/config.yaml
Comment (jwflory) [reply exampleSite/config.yaml]: The HTML validation issues were an issue with my environment. I tried again and the errors cleared. Not sure what happened the first time.

I did a manual rebase now and squashed the commits down to a single commit. This one was especially tricky. It will be worthwhile to review git rebasing together in terms of workflows. But I think the circumstances were unusual here since the change was pending for some time, and other changes entered the main code base.

**Review by jwflory — APPROVED** [review]
This looks ready to me. I force-pushed a manual `git rebase` on top of the `main` branch, which cleaned up the commit history. The situation was more complicated than normal. Note that you will need to run the following commands to get a working copy of the git branch again:

```
git checkout edit-this-page-btn
git fetch origin edit-this-page-btn
git reset --hard edit-this-page-btn
```

This looks ready to merge, although before merging, I want to validate that the escaping challenge was resolved. It escapes correctly for me in Firefox 100. Is this true for you as well?

**Comment by Neha9849** [comment]
To avoid all this mess. I did a hard `reset` to make this my branch even with this repo main branch. Because of which this PR got automatically closed.
I have created a new PR #110 with the exact changes.

**Comment by jwflory** [comment]
No problem. Let's review #110 instead. 👍🏻 

> _**@Neha9849 wrote**_…
> I accidently added some commits in my main branch and which resulted in all this mess. I won't repeat this mistake, Sorry🥲 

It happens! Playing with git history and rebasing commits gets messy. There will be more opportunities to practice and improve with git 💪🏻😎

**Pull request closed without merging**
